File: algorithms/sorting/mergesort.rb
def merge_sort(arr)
    if arr.size <= 1
        return arr
    end
    size = arr.size
    middle = size / 2
   
    left = arr[0...middle]
    right = arr[middle...size]
    # OR 
    # left = arr.take(middle)
    # right = arr.drop(middle)

    left_sorted = merge_sort(left)
    right_sorted = merge_sort(right)

    return merge(left_sorted, right_sorted)
end

def merge(left, right)
    results = []
    while left.length > 0 && right.length > 0
        if left[0] < right[0]
            results << left.shift
        else
            results << right.shift
        end
    end

    results += left if left.length > 0
    results += right if right.length > 0
    results
end

arr = [9,7,3,4,1,8,2]
p merge_sort(arr)

File: algorithms/sorting/merge_sort.py
def merge_sort(items):
  if len(items) <= 1:
    return items

  middle_index = len(items) // 2
  left_split = items[:middle_index]
  right_split = items[middle_index:]

  left_sorted = merge_sort(left_split)
  right_sorted = merge_sort(right_split)

  return merge(left_sorted, right_sorted)

def merge(left, right):
  result = []
  print(left, right)
  while (left and right):
    if left[0] < right[0]:
      result.append(left.pop(0))
    else:
      result.append(right.pop(0))

  if left:
    result += left
  if right:
    result += right

  return result

# Time & Space
# time: O(n log n)
# space: O(n)

unordered_list = [356, 493, 746, 264, 569, 949, 895, 125]
ordered_list = merge_sort(unordered_list)
print('ordered_list', ordered_list)
